key.go

```diff
@@ -0,0 +1,806 @@
+	KeyOpSign KeyOp = 1
+
+	// The key is used for verification of signatures.
+	KeyOpVerify KeyOp = 2
+
+	// The key is used for key transport encryption.
+	KeyOpEncrypt KeyOp = 3
+
+	// The key is used for key transport decryption. Requires private key fields.
+	KeyOpDecrypt KeyOp = 4
+
+	// The key is used for key wrap encryption.
+	KeyOpWrapKey KeyOp = 5
+
+	// The key is used for key wrap decryption.
+	KeyOpUnwrapKey KeyOp = 6
+
+	// The key is used for deriving keys. Requires private key fields.
+	KeyOpDeriveKey KeyOp = 7
+
+	// The key is used for deriving bits not to be used as a key. Requires
+	// private key fields.
+	KeyOpDeriveBits KeyOp = 8
+
+	// The key is used for creating MACs.
+	KeyOpMACCreate KeyOp = 9
+
+	// The key is used for validating MACs.
+	KeyOpMACVerify KeyOp = 10
+)
+
+// KeyOp represents a key_ops value used to restrict purposes for which a Key
+// may be used.
+type KeyOp int64
+
+// KeyOpFromString returns the KeyOp corresponding to the specified name.
+// The values are taken from https://www.rfc-editor.org/rfc/rfc7517#section-4.3
+func KeyOpFromString(val string) KeyOp {
+	switch val {
+	case "sign":
+		return KeyOpSign
+	case "verify":
+		return KeyOpVerify
+	case "encrypt":
+		return KeyOpEncrypt
+	case "decrypt":
+		return KeyOpDecrypt
+	case "wrapKey":
+		return KeyOpWrapKey
+	case "unwrapKey":
+		return KeyOpUnwrapKey
+	case "deriveKey":
+		return KeyOpDeriveKey
+	case "deriveBits":
+		return KeyOpDeriveBits
+	default:
+		return KeyOpInvalid
+	}
+}
+
+func (ko KeyOp) String() string {
+	switch ko {
+	case KeyOpSign:
+		return "sign"
+	case KeyOpVerify:
+		return "verify"
+	case KeyOpEncrypt:
+		return "encrypt"
+	case KeyOpDecrypt:
+		return "decrypt"
+	case KeyOpWrapKey:
+		return "wrapKey"
+	case KeyOpUnwrapKey:
+		return "unwrapKey"
+	case KeyOpDeriveKey:
+		return "deriveKey"
+	case KeyOpDeriveBits:
+		return "deriveBits"
+	case KeyOpMACCreate:
+		return "MAC create"
+	case KeyOpMACVerify:
+		return "MAC verify"
+	default:
+		return "unknown key_op value " + strconv.Itoa(int(ko))
+	}
+}
+
+func (ko KeyOp) IsSupported() bool {
+	return ko >= 1 && ko <= 10
+}
+
+// MarshalCBOR marshals the KeyOp as a CBOR int.
+func (ko KeyOp) MarshalCBOR() ([]byte, error) {
+	return encMode.Marshal(int64(ko))
+}
+
+// UnmarshalCBOR populates the KeyOp from the provided CBOR value (must be int
+// or tstr).
+func (ko *KeyOp) UnmarshalCBOR(data []byte) error {
+	var raw intOrStr
+
+	if err := raw.UnmarshalCBOR(data); err != nil {
+		return fmt.Errorf("invalid key_ops value %w", err)
+	}
+
+	if raw.IsString() {
+		v := KeyOpFromString(raw.String())
+		if v == KeyOpInvalid {
+			return fmt.Errorf("unknown key_ops value %q", raw.String())
+		}
+
+		*ko = v
+	} else {
+		v := raw.Int()
+		*ko = KeyOp(v)
+
+		if !ko.IsSupported() {
+			return fmt.Errorf("unknown key_ops value %d", v)
+		}
+	}
+
+	return nil
+}
+
+const (
+	// Octet Key Pair
+	KeyTypeOkp KeyType = 1
```

Comment:
Is KeyType = 0 a unused  reserved value? If so, we can use it to denotate an invalid key type in the same way we have `KeyOpInvalid` and `AlgorithmInvalid`.